### PR TITLE
Add nim linter

### PR DESCRIPTION
**Pull request by Jipok**

### CODE REVIEW

File: language_nim.lua [hunk 1]
```diff
@@ -0,0 +1,11 @@
+local config = require "core.config"
+local linter = require "plugins.linter"
+
+config.nimcheck_args = {}
+
+linter.add_language {
+  file_patterns = {"%.nim$"},
+  warning_pattern = "[^:]%((%d+), (%d+)%) ([^\n]+)",
+  command = "nim check $ARGS $FILENAME 2>&1",
+  args = config.luacheck_args
```
Comment (drmargarido) [inline language_nim.lua]: You created the nimcheck_args variable in the configuration but passed the config.luacheck_args when adding the language.

Comment (Jipok) [reply language_nim.lua]: Not only that, but also the wrong file name. Which I copied and destroyed my nim syntax highlighting work which I did 2 hours(

**Comment by Jipok** [comment]
Sometimes a different file may be specified in the logs. For example, now I called the linter for tmp.nim:
```nim
/home/kiv/dev/lite/tmp.nim(276, 17) template/generic instantiation of `[]=` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/accessors_macros_write.nim(51, 12) template/generic instantiation of `[]` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/accessors_macros_read.nim(55, 12) template/generic instantiation of `slice_typed_dispatch` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_accessors_macros_read.nim(63, 13) template/generic instantiation of `slicerImpl` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_accessors_macros_read.nim(29, 25) template/generic instantiation of `newException` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_accessors_macros_read.nim(30, 9) Warning: See corresponding Defect; IndexError is deprecated [Deprecated]
/home/kiv/dev/lite/tmp.nim(276, 17) template/generic instantiation of `[]=` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/accessors_macros_write.nim(51, 12) template/generic instantiation of `slice_typed_dispatch_mut` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_accessors_macros_write.nim(150, 18) template/generic instantiation of `slicerMutImpl_T` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_accessors_macros_write.nim(143, 16) template/generic instantiation of `check_shape` from here
/home/kiv/.nimble/pkgs/arraymancer-0.5.2/tensor/private/p_checks.nim(90, 24) Warning: See corresponding Defect; IndexError is deprecated [Deprecated]
/home/kiv/dev/lite/tmp.nim(387, 18) template/generic instantiation of `newShakespeareNet` from here
/home/kiv/dev/lite/tmp.nim(196, 45) template/generic instantiation of `[]=` from here
```
How can I exclude extra lines?

**Comment by drmargarido** [comment]
I don't use nim, but usually the linters have arguments which allow you to filter some of the errors/warnings. I don't know if these work with the check option but you could try some - https://nim-lang.org/docs/nimc.html

**Comment by Jipok** [comment]
Unfortunately, this cannot be done through arguments, but I did it through grep.
By the way, and my first version, and current one most likely will not work on windows

**Comment by drmargarido** [comment]
Would be better to keep the linters working in windows and linux if possible. 
Also, this kind of problem also happens in the golang linters since they work in whole packages.

Instead of doing the filtering with a grep we could extend the match pattern to allow the placement of a $FILENAME value there, so we could do that filtering directly in the lua pattern.

I will do an implementation to allow that and then come back to this pull request.

**Comment by Jipok** [comment]
match_pattern in lua does not support lookahead as far as I know, so i used grep. I have no idea how to do without it.
Windows support is not related to using grep. The problem is that I do not know how to check the beginning of the path for Windows and Unix at the same time. I am currently checking `/`. Ideally, I need to check `(/|[%a]:\\)`, but it seems that Lua does not support this. 

By the way, nim in addition to warnings, it also displays errors and hints. But, due to restrictions, they are displayed the same.

**Comment by drmargarido** [comment]
I allowed the usage of the $FILENAME in the warning_pattern. Check the commit -> ed432a9
I modified the linter you created to not use grep and to use a negate capture [^\//] instead of trying to looking ahead.
Also the nim check linter reported multiple times the same warning so I added some de-duplication to them.

This way it should work in linux and windows. Also since we are not passing the results in the command line between multiple commands, the editor time blocked while waiting for the command result is also smaller.

I downloaded a nim project from github to validate that it is working as wanted, but would be great if you could also test it and check if everything is working as expected.

I still need to escape the $FILENAME when is put into the pattern since right now it can inject patterns in there if they are in the file path.

**Comment by drmargarido** [comment]
I did a pull request with the complete implementation in #16 . 

While the implementation is new, it was started by your implementation and initial investigation. Thanks for the PR, if you have any feedback or see something missing comment in the other PR.

**Pull request closed without merging**